{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9843c7b8-185f-40e9-b95e-b13fc47bf361",
   "metadata": {},
   "outputs": [],
   "source": [
    "import ijson\n",
    "genes_matched = []\n",
    "with open('.\\data\\steps data output\\step-3\\3_matched_biological_entites', 'r', encoding='utf-8') as f:\n",
    "    # Parse items one by one (adjust 'item' based on your JSON structure)\n",
    "    for item in ijson.items(f, 'item'):\n",
    "        genes_matched.append(item)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f8335bb8-64a4-498d-999e-c1e78c53f0ff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2120"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(genes_matched)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "42d33319-d132-4376-a050-e3f93ada6f12",
   "metadata": {},
   "outputs": [],
   "source": [
    "import ijson\n",
    "df_genes = []\n",
    "with open('.\\data\\steps data output\\step-2\\3_biological_entities', 'r', encoding='utf-8') as f:\n",
    "    # Parse items one by one (adjust 'item' based on your JSON structure)\n",
    "    for item in ijson.items(f, 'item'):\n",
    "        df_genes.append(item)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5a7f35ec-4fb0-4d78-96f5-18f90e9e623a",
   "metadata": {},
   "outputs": [],
   "source": [
    "Gene_relations = {}\n",
    "unique_key = lambda x: (x['identifier'], x['pmid'], x['type'], x['word'], x['sentence'])\n",
    "\n",
    "for i in genes_matched:\n",
    "    for x in df_genes:\n",
    "        if i['Identifier'].lower() == x['identifier'].lower():\n",
    "            item = {\n",
    "                'identifier': x['identifier'],\n",
    "                'pmid': x['PMID'],\n",
    "                'type': x['type'],\n",
    "                'word': x['word'],\n",
    "                'sentence': x['sentence']\n",
    "            }\n",
    "            key = unique_key(item)\n",
    "            if key not in Gene_relations:\n",
    "                Gene_relations[key] = item\n",
    "\n",
    "# Convert back to list if needed\n",
    "Gene_relations = list(Gene_relations.values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7c6cba9c-5388-4643-98a9-ca6842688f30",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "11771"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(Gene_relations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "6aceb7c6-ca22-4169-9c98-d34c33cbdc21",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "\n",
    "# Group gene and disease relations by sentence\n",
    "genes_by_sentence_1 = defaultdict(list)\n",
    "genes_by_sentence_2 = defaultdict(list)\n",
    "\n",
    "for gene1 in Gene_relations:\n",
    "    genes_by_sentence_1[gene1['sentence']].append(gene1)\n",
    "\n",
    "for gene2 in Gene_relations:\n",
    "    genes_by_sentence_2[gene2['sentence']].append(gene2)\n",
    "\n",
    "# Create unique pairs\n",
    "seen_keys = set()\n",
    "Gene_gene_relations = []\n",
    "\n",
    "for sentence in genes_by_sentence_2:\n",
    "    if sentence in genes_by_sentence_1:\n",
    "        for gene2 in genes_by_sentence_2[sentence]:\n",
    "            for gene1 in genes_by_sentence_1[sentence]:\n",
    "                key1 = (gene2['pmid'], sentence, gene2['identifier'], gene1['identifier'])\n",
    "                key2 = (gene2['pmid'], sentence, gene1['identifier'], gene2['identifier'])\n",
    "                if key1 not in seen_keys and key2 not in seen_keys and gene2['identifier']!= gene1['identifier']:\n",
    "                    seen_keys.add(key1)\n",
    "                    seen_keys.add(key2)\n",
    "                    Gene_gene_relations.append({\n",
    "                        'pmid': gene2['pmid'],\n",
    "                        'sentence': sentence,\n",
    "                        'target1_identifier': gene2['identifier'],\n",
    "                        'target1_type': gene2['type'],\n",
    "                        'target1_word': gene2['word'],\n",
    "                        'target2_identifier': gene1['identifier'],\n",
    "                        'target2_type': gene1['type'],\n",
    "                        'target2_word': gene1['word']\n",
    "                    })\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "ffee73b6-078d-4407-a4e1-260f09588faf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8724"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(Gene_gene_relations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18db0782-31e8-4a39-a4cc-32e524a07944",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "with open('.\\data\\steps data output\\step-4\\gene_gene_relations.json', \"w\") as json_file:\n",
    "    json.dump(Gene_gene_relations, json_file, indent=4)  # indent=4 for pretty-printing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "855008f5-ccc9-4579-a781-932ee66688ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "with open('.\\data\\steps data output\\step-4\\relation_genes.json', \"w\") as json_file:\n",
    "    json.dump(Gene_relations, json_file, indent=4)  # indent=4 for pretty-printing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e438ce46-4ce2-48cf-9c00-c2f14f833b41",
   "metadata": {},
   "outputs": [],
   "source": [
    "import ijson\n",
    "diseases_matched = []\n",
    "with open('.\\data\\steps data output\\step-3\\3_matched_biological_entites\\matched_diseases.json', 'r', encoding='utf-8') as f:\n",
    "    # Parse items one by one (adjust 'item' based on your JSON structure)\n",
    "    for item in ijson.items(f, 'item'):\n",
    "        diseases_matched.append(item)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "56409e9b-3cc1-4c62-9ce6-1e68d946eb2c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "15472"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(diseases_matched)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d02bab83-c8fc-45b4-9d2c-1fec34d4ab18",
   "metadata": {},
   "outputs": [],
   "source": [
    "import ijson\n",
    "df_diseases = []\n",
    "with open('.\\data\\steps data output\\step-2\\3_biological_entities\\Disease_Extraction.json', 'r', encoding='utf-8') as f:\n",
    "    # Parse items one by one (adjust 'item' based on your JSON structure)\n",
    "    for item in ijson.items(f, 'item'):\n",
    "        df_diseases.append(item)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "dcd95026-9a8e-481d-8322-2cc5ece7732f",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "\n",
    "# Preprocess df_diseases into a lookup dictionary with uppercased identifiers\n",
    "df_diseases_lookup = defaultdict(list)\n",
    "for x in df_diseases:\n",
    "    df_diseases_lookup[x['identifier']].append(x)\n",
    "\n",
    "# Use a set to track seen keys and build Disease_relations list\n",
    "unique_key = lambda x: (x['identifier'], x['pmid'], x['type'], x['word'], x['sentence'])\n",
    "seen_keys = set()\n",
    "Disease_relations = []\n",
    "\n",
    "for i in diseases_matched:\n",
    "    matches = df_diseases_lookup.get(i['Identifier'].upper(), [])\n",
    "    for x in matches:\n",
    "        item = {\n",
    "            'identifier': x['identifier'],\n",
    "            'pmid': x['PMID'],\n",
    "            'type': x['type'],\n",
    "            'word': x['word'],\n",
    "            'sentence': x['sentence']\n",
    "        }\n",
    "        key = unique_key(item)\n",
    "        if key not in seen_keys:\n",
    "            seen_keys.add(key)\n",
    "            Disease_relations.append(item)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "51091f10-1031-4e62-93eb-f22587c94ee6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "187187"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(Disease_relations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "e539c804-6ef5-422b-bf82-101251b69cbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "\n",
    "# Group gene and disease relations by sentence\n",
    "diseases_by_sentence_1 = defaultdict(list)\n",
    "diseases_by_sentence_2 = defaultdict(list)\n",
    "\n",
    "for disease1 in Disease_relations:\n",
    "    diseases_by_sentence_1[disease1['sentence']].append(disease1)\n",
    "\n",
    "for disease2 in Disease_relations:\n",
    "    diseases_by_sentence_2[disease2['sentence']].append(disease2)\n",
    "\n",
    "# Create unique pairs\n",
    "seen_keys = set()\n",
    "Disease_disease_relations = []\n",
    "\n",
    "for sentence in diseases_by_sentence_2:\n",
    "    if sentence in diseases_by_sentence_1:\n",
    "        for disease2 in diseases_by_sentence_2[sentence]:\n",
    "            for disease1 in diseases_by_sentence_1[sentence]:\n",
    "                key1 = (disease2['pmid'], sentence, disease2['identifier'], disease1['identifier'])\n",
    "                key2 = (disease2['pmid'], sentence, disease1['identifier'], disease2['identifier'])\n",
    "                if key1 not in seen_keys and key2 not in seen_keys and disease2['identifier']!= disease1['identifier']:\n",
    "                    seen_keys.add(key1)\n",
    "                    seen_keys.add(key2)\n",
    "                    Disease_disease_relations.append({\n",
    "                        'pmid': disease2['pmid'],\n",
    "                        'sentence': sentence,\n",
    "                        'target1_identifier': disease2['identifier'],\n",
    "                        'target1_type': disease2['type'],\n",
    "                        'target1_word': disease2['word'],\n",
    "                        'target2_identifier': disease1['identifier'],\n",
    "                        'target2_type': disease1['type'],\n",
    "                        'target2_word': disease1['word']\n",
    "                    })\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "a6e6388c-6ac2-4f09-a7f3-22deb3818399",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "72225"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(Disease_disease_relations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "01c9bfb6-0152-43cc-a432-fb3d2da8b075",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'pmid': '24903423',\n",
       " 'sentence': 'These tumors may arise sporadically or be associated to various syndromes, namely multiple endocrine neoplasia type 2, neurofibromatosis type 1, Von Hippel-Lindau syndrome, and hereditary paraganglioma-pheochromocytoma syndromes.',\n",
       " 'target1_identifier': 'MESH:D009456',\n",
       " 'target1_type': 'Disease',\n",
       " 'target1_word': 'neurofibromatosis type 1',\n",
       " 'target2_identifier': 'MESH:D006623',\n",
       " 'target2_type': 'Disease',\n",
       " 'target2_word': 'Von Hippel-Lindau syndrome'}"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Disease_disease_relations[2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cd5d317a-11a1-4c1f-ad8b-a736bac4bc1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "with open('.\\data\\steps data output\\step-4\\disease_disease_relations.json', \"w\") as json_file:\n",
    "    json.dump(Disease_disease_relations, json_file, indent=4)  # indent=4 for pretty-printing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a4f7f34-18c5-4709-a4f7-a193b85cacc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "with open('.\\data\\steps data output\\step-4\\relation_diseases.json', \"w\") as json_file:\n",
    "    json.dump(Disease_relations, json_file, indent=4)  # indent=4 for pretty-printing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "91d3aa2a-dddd-42ae-a5cc-c9dfd55a4b6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "\n",
    "# Group gene and disease relations by sentence\n",
    "genes_by_sentence = defaultdict(list)\n",
    "diseases_by_sentence = defaultdict(list)\n",
    "\n",
    "for gene in Gene_relations:\n",
    "    genes_by_sentence[gene['sentence']].append(gene)\n",
    "\n",
    "for disease in Disease_relations:\n",
    "    diseases_by_sentence[disease['sentence']].append(disease)\n",
    "\n",
    "# Create unique pairs\n",
    "seen_keys = set()\n",
    "Disease_gene_relations = []\n",
    "\n",
    "for sentence in diseases_by_sentence:\n",
    "    if sentence in genes_by_sentence:\n",
    "        for disease in diseases_by_sentence[sentence]:\n",
    "            for gene in genes_by_sentence[sentence]:\n",
    "                key = (disease['pmid'], sentence, disease['identifier'], gene['identifier'])\n",
    "                if key not in seen_keys:\n",
    "                    seen_keys.add(key)\n",
    "                    Disease_gene_relations.append({\n",
    "                        'pmid': disease['pmid'],\n",
    "                        'sentence': sentence,\n",
    "                        'target1_identifier': disease['identifier'],\n",
    "                        'target1_type': disease['type'],\n",
    "                        'target1_word': disease['word'],\n",
    "                        'target2_identifier': gene['identifier'],\n",
    "                        'target2_type': gene['type'],\n",
    "                        'target2_word': gene['word']\n",
    "                    })\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "bf0a6154-b381-4860-aea0-35a0577f8036",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6063"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(Disease_gene_relations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ea9f1dd9-083a-4a67-8ad7-f36967119f04",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "with open('.\\data\\steps data output\\step-4\\disease_gene_relations.json', \"w\") as json_file:\n",
    "    json.dump(Disease_gene_relations, json_file, indent=4)  # indent=4 for pretty-printing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "256be58f-9633-4928-baa0-7a323eb796a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import ijson\n",
    "df_chemicals = []\n",
    "with open('.\\data\\steps data output\\step-2\\3_biological_entities\\Chemical_Extraction.json', 'r', encoding='utf-8') as f:\n",
    "    # Parse items one by one (adjust 'item' based on your JSON structure)\n",
    "    for item in ijson.items(f, 'item'):\n",
    "        df_chemicals.append(item)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "7f3f3ee5-7a24-4e14-8728-fe0c790cae08",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "\n",
    "# Group chemical and disease relations by sentence\n",
    "chemicals_by_sentence = defaultdict(list)\n",
    "diseases_by_sentence = defaultdict(list)\n",
    "\n",
    "for chemical in df_chemicals:\n",
    "    chemicals_by_sentence[chemical['sentence']].append(chemical)\n",
    "\n",
    "for disease in Disease_relations:\n",
    "    diseases_by_sentence[disease['sentence']].append(disease)\n",
    "\n",
    "# Create unique pairs\n",
    "seen_keys = set()\n",
    "Disease_chemical_relations = []\n",
    "\n",
    "for sentence in diseases_by_sentence:\n",
    "    if sentence in chemicals_by_sentence:\n",
    "        for disease in diseases_by_sentence[sentence]:\n",
    "            for chemical in chemicals_by_sentence[sentence]:\n",
    "                key = (disease['pmid'], sentence, disease['identifier'], chemical['identifier'])\n",
    "                if key not in seen_keys:\n",
    "                    seen_keys.add(key)\n",
    "                    Disease_chemical_relations.append({\n",
    "                        'pmid': disease['pmid'],\n",
    "                        'sentence': sentence,\n",
    "                        'target1_identifier': disease['identifier'],\n",
    "                        'target1_type': disease['type'],\n",
    "                        'target1_word': disease['word'],\n",
    "                        'target2_identifier': chemical['identifier'],\n",
    "                        'target2_type': chemical['type'],\n",
    "                        'target2_word': chemical['word']\n",
    "                    })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "89fc1d60-a211-4910-a556-f6570da69332",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "13598"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(Disease_chemical_relations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "202f8451-8dd6-4f6a-8860-21e1b4160c08",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'pmid': '25588595',\n",
       " 'sentence': 'BACKGROUND: In nonvalvular atrial fibrillation (NVAF), rivaroxaban is used to prevent stroke and systemic embolism.',\n",
       " 'target1_identifier': 'MESH:D004617',\n",
       " 'target1_type': 'Disease',\n",
       " 'target1_word': 'systemic embolism',\n",
       " 'target2_identifier': 'MESH:D000069552',\n",
       " 'target2_type': 'Chemical',\n",
       " 'target2_word': 'rivaroxaban'}"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Disease_chemical_relations[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "808af1f5-7066-464f-a69c-0a8b1abdc33e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "with open('.\\data\\steps data output\\step-4\\disease_chemical_relations.json', \"w\") as json_file:\n",
    "    json.dump(Disease_chemical_relations, json_file, indent=4)  # indent=4 for pretty-printing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c7a8252-d27e-4c1c-94de-98923cd47da7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import ijson\n",
    "df_chromosomes = []\n",
    "with open('.\\data\\steps data output\\step-2\\3_biological_entities\\Chromosome_Extraction.json', 'r', encoding='utf-8') as f:\n",
    "    # Parse items one by one (adjust 'item' based on your JSON structure)\n",
    "    for item in ijson.items(f, 'item'):\n",
    "        df_chromosomes.append(item)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "09ff10c7-8f59-481d-89b4-b3d342fd3cd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "\n",
    "# Group chromosome and disease relations by sentence\n",
    "chromosomes_by_sentence = defaultdict(list)\n",
    "diseases_by_sentence = defaultdict(list)\n",
    "\n",
    "for chromosome in df_chromosomes:\n",
    "    chromosomes_by_sentence[chromosome['sentence']].append(chromosome)\n",
    "\n",
    "for disease in Disease_relations:\n",
    "    diseases_by_sentence[disease['sentence']].append(disease)\n",
    "\n",
    "# Create unique pairs\n",
    "seen_keys = set()\n",
    "Disease_chromosome_relations = []\n",
    "\n",
    "for sentence in diseases_by_sentence:\n",
    "    if sentence in chromosomes_by_sentence:\n",
    "        for disease in diseases_by_sentence[sentence]:\n",
    "            for chromosome in chromosomes_by_sentence[sentence]:\n",
    "                key = (disease['pmid'], sentence, disease['identifier'], chromosome['identifier'])\n",
    "                if key not in seen_keys:\n",
    "                    seen_keys.add(key)\n",
    "                    Disease_chromosome_relations.append({\n",
    "                        'pmid': disease['pmid'],\n",
    "                        'sentence': sentence,\n",
    "                        'target1_identifier': disease['identifier'],\n",
    "                        'target1_type': disease['type'],\n",
    "                        'target1_word': disease['word'],\n",
    "                        'target2_identifier': chromosome['identifier'],\n",
    "                        'target2_type': chromosome['type'],\n",
    "                        'target2_word': chromosome['word']\n",
    "                    })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "9777288d-9257-41a8-96ed-785e2468f875",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "77"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(Disease_chromosome_relations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a90e791b-a6cf-4d88-bbbb-a1e1ede3542c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "with open('.\\data\\steps data output\\step-4\\disease_chromosome_relations.json', \"w\") as json_file:\n",
    "    json.dump(Disease_chromosome_relations, json_file, indent=4)  # indent=4 for pretty-printing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "47cbace8-12a2-4a8f-a836-008639189b89",
   "metadata": {},
   "outputs": [],
   "source": [
    "import ijson\n",
    "df_variants = []\n",
    "with open('.\\data\\steps data output\\step-2\\3_biological_entities\\Variant_Extraction.json', 'r', encoding='utf-8') as f:\n",
    "    # Parse items one by one (adjust 'item' based on your JSON structure)\n",
    "    for item in ijson.items(f, 'item'):\n",
    "        df_variants.append(item)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "f7e4993e-33bc-4b1c-9f03-3558335bada6",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "\n",
    "# Group variant and disease relations by sentence\n",
    "variants_by_sentence = defaultdict(list)\n",
    "diseases_by_sentence = defaultdict(list)\n",
    "\n",
    "for variant in df_variants:\n",
    "    variants_by_sentence[variant['sentence']].append(variant)\n",
    "\n",
    "for disease in Disease_relations:\n",
    "    diseases_by_sentence[disease['sentence']].append(disease)\n",
    "\n",
    "# Create unique pairs\n",
    "seen_keys = set()\n",
    "Disease_variant_relations = []\n",
    "\n",
    "for sentence in diseases_by_sentence:\n",
    "    if sentence in variants_by_sentence:\n",
    "        for disease in diseases_by_sentence[sentence]:\n",
    "            for variant in variants_by_sentence[sentence]:\n",
    "                key = (disease['pmid'], sentence, disease['identifier'], variant['identifier'])\n",
    "                if key not in seen_keys:\n",
    "                    seen_keys.add(key)\n",
    "                    Disease_variant_relations.append({\n",
    "                        'pmid': disease['pmid'],\n",
    "                        'sentence': sentence,\n",
    "                        'target1_identifier': disease['identifier'],\n",
    "                        'target1_type': disease['type'],\n",
    "                        'target1_word': disease['word'],\n",
    "                        'target2_identifier': variant['identifier'],\n",
    "                        'target2_type': variant['type'],\n",
    "                        'target2_word': variant['word']\n",
    "                    })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "8520d2af-e4e6-42b3-b165-b391b8d4fc5f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "264"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(Disease_variant_relations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9bd3e8f4-bba6-42fd-9326-12592e06663a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "with open('.\\data\\steps data output\\step-4\\disease_variant_relations.json', \"w\") as json_file:\n",
    "    json.dump(Disease_variant_relations, json_file, indent=4)  # indent=4 for pretty-printing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a45108e4-a88c-46a0-9ee4-b02d3a242cca",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "\n",
    "# Group chromosome and genes relations by sentence\n",
    "chromosomes_by_sentence = defaultdict(list)\n",
    "genes_by_sentence = defaultdict(list)\n",
    "\n",
    "for chromosome in df_chromosomes:\n",
    "    chromosomes_by_sentence[chromosome['sentence']].append(chromosome)\n",
    "\n",
    "for gene in Gene_relations:\n",
    "    genes_by_sentence[gene['sentence']].append(gene)\n",
    "\n",
    "# Create unique pairs\n",
    "seen_keys = set()\n",
    "Gene_chromosome_relations = []\n",
    "\n",
    "for sentence in genes_by_sentence:\n",
    "    if sentence in chromosomes_by_sentence:\n",
    "        for gene in genes_by_sentence[sentence]:\n",
    "            for chromosome in chromosomes_by_sentence[sentence]:\n",
    "                key = (gene['pmid'], sentence, gene['identifier'], chromosome['identifier'])\n",
    "                if key not in seen_keys:\n",
    "                    seen_keys.add(key)\n",
    "                    Gene_chromosome_relations.append({\n",
    "                        'pmid': gene['pmid'],\n",
    "                        'sentence': sentence,\n",
    "                        'target1_identifier': gene['identifier'],\n",
    "                        'target1_type': gene['type'],\n",
    "                        'target1_word': gene['word'],\n",
    "                        'target2_identifier': chromosome['identifier'],\n",
    "                        'target2_type': chromosome['type'],\n",
    "                        'target2_word': chromosome['word']\n",
    "                    })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "89ce45dd-ac12-461e-9ac9-c9b48f3de9b7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "194"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(Gene_chromosome_relations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "2a38deb3-f557-4aad-923e-c9ea9c54fb05",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'pmid': '31706190',\n",
       " 'sentence': 'This study investigated whether any of the six initially discovered genomic loci associating with RLS (BTBD9, MEIS1, PTPRD, MAP2K5/SKOR1, TOX3, and an intergenic region on chromosome 2), were more strongly associated with complaints of painful versus non-painful RLS.',\n",
       " 'target1_identifier': '5607',\n",
       " 'target1_type': 'Gene',\n",
       " 'target1_word': 'MAP2K5',\n",
       " 'target2_identifier': '2',\n",
       " 'target2_type': 'Chromosome',\n",
       " 'target2_word': 'chromosome 2'}"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Gene_chromosome_relations[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abe56068-9c62-47ef-9d29-1130c224fcd9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "with open('.\\data\\steps data output\\step-4\\gene_chromosome_relations.json', \"w\") as json_file:\n",
    "    json.dump(Gene_chromosome_relations, json_file, indent=4)  # indent=4 for pretty-printing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "23631a9e-dbd8-4afd-9cb1-2f7b491877c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "\n",
    "# Group chemical and disease relations by sentence\n",
    "chemicals_by_sentence = defaultdict(list)\n",
    "genes_by_sentence = defaultdict(list)\n",
    "\n",
    "for chemical in df_chemicals:\n",
    "    chemicals_by_sentence[chemical['sentence']].append(chemical)\n",
    "\n",
    "for gene in Gene_relations:\n",
    "    genes_by_sentence[gene['sentence']].append(gene)\n",
    "\n",
    "# Create unique pairs\n",
    "seen_keys = set()\n",
    "Gene_chemical_relations = []\n",
    "\n",
    "for sentence in genes_by_sentence:\n",
    "    if sentence in chemicals_by_sentence:\n",
    "        for gene in genes_by_sentence[sentence]:\n",
    "            for chemical in chemicals_by_sentence[sentence]:\n",
    "                key = (disease['pmid'], sentence, gene['identifier'], chemical['identifier'])\n",
    "                if key not in seen_keys:\n",
    "                    seen_keys.add(key)\n",
    "                    Gene_chemical_relations.append({\n",
    "                        'pmid': gene['pmid'],\n",
    "                        'sentence': sentence,\n",
    "                        'target1_identifier': gene['identifier'],\n",
    "                        'target1_type': gene['type'],\n",
    "                        'target1_word': gene['word'],\n",
    "                        'target2_identifier': chemical['identifier'],\n",
    "                        'target2_type': chemical['type'],\n",
    "                        'target2_word': chemical['word']\n",
    "                    })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "5e7fddbc-36b4-4a0b-839e-90a69022a398",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1908"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(Gene_chemical_relations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "f80d455b-c9b7-4bbc-8a51-4ad9092375dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'pmid': '32393603',\n",
       " 'sentence': 'PURPOSE: Differentiation syndrome (DS) is a serious adverse reaction of isocitrate dehydrogenase (IDH) inhibitors ivosidenib and enasidenib in patients with (IDH)1- and IDH2-mutated acute myeloid leukemia (AML), respectively.',\n",
       " 'target1_identifier': '3417',\n",
       " 'target1_type': 'Gene',\n",
       " 'target1_word': 'isocitrate dehydrogenase',\n",
       " 'target2_identifier': 'MESH:C000627630',\n",
       " 'target2_type': 'Chemical',\n",
       " 'target2_word': 'ivosidenib'}"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Gene_chemical_relations[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98524939-5df9-45c6-b618-c0ceeb5f8f1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "with open('.\\data\\steps data output\\step-4\\gene_chemical_relations.json', \"w\") as json_file:\n",
    "    json.dump(Gene_chemical_relations, json_file, indent=4)  # indent=4 for pretty-printing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "bf4747be-8221-4ce6-9c23-63d65afe0eb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "\n",
    "# Group variant and disease relations by sentence\n",
    "variants_by_sentence = defaultdict(list)\n",
    "genes_by_sentence = defaultdict(list)\n",
    "\n",
    "for variant in df_variants:\n",
    "    variants_by_sentence[variant['sentence']].append(variant)\n",
    "\n",
    "for gene in Gene_relations:\n",
    "    genes_by_sentence[gene['sentence']].append(gene)\n",
    "\n",
    "# Create unique pairs\n",
    "seen_keys = set()\n",
    "Gene_variant_relations = []\n",
    "\n",
    "for sentence in diseases_by_sentence:\n",
    "    if sentence in variants_by_sentence:\n",
    "        for gene in genes_by_sentence[sentence]:\n",
    "            for variant in variants_by_sentence[sentence]:\n",
    "                key = (gene['pmid'], sentence, gene['identifier'], variant['identifier'])\n",
    "                if key not in seen_keys:\n",
    "                    seen_keys.add(key)\n",
    "                    Gene_variant_relations.append({\n",
    "                        'pmid': gene['pmid'],\n",
    "                        'sentence': sentence,\n",
    "                        'target1_identifier': gene['identifier'],\n",
    "                        'target1_type': gene['type'],\n",
    "                        'target1_word': gene['word'],\n",
    "                        'target2_identifier': variant['identifier'],\n",
    "                        'target2_type': variant['type'],\n",
    "                        'target2_word': variant['word']\n",
    "                    })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "82d64436-1cb3-4d1b-89b3-8387d9d35007",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "172"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(Gene_variant_relations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "5a91a85c-d395-4ef7-9bbe-d3761174fa8b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'pmid': '33764904',\n",
       " 'sentence': 'Compartment-specific gene expression data and immunohistochemistry were also queried.RESULTSThe identical germline mutation of the PALLD gene (NM_001166108.1:c.G154A:p.D52N) was detected in the index patient with pancreatic cancer and the tumor tissue of her sister.',\n",
       " 'target1_identifier': '23022',\n",
       " 'target1_type': 'Gene',\n",
       " 'target1_word': 'PALLD',\n",
       " 'target2_identifier': 'tmVar:c|SUB|G|154|A;HGVS:c.154G>A;VariantGroup:1;CorrespondingGene:23022;RS#:539876255;CorrespondingSpecies:9606;CA#:3135302',\n",
       " 'target2_type': 'Variant',\n",
       " 'target2_word': 'c.G154A'}"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Gene_variant_relations[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa40e743-7da6-4d3e-8c19-c8e7157416cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "with open('.\\data\\steps data output\\step-4\\gene_variant_relations.json', \"w\") as json_file:\n",
    "    json.dump(Gene_variant_relations, json_file, indent=4)  # indent=4 for pretty-printing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ce89ec1-0c4b-47b3-8b36-310c69fd34fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "from transformers import AutoTokenizer, AutoModelForSeq2SeqLM\n",
    "from neo4j import GraphDatabase\n",
    "from tqdm import tqdm\n",
    "\n",
    "# === Load REBEL model once ===\n",
    "print(\"üîÑ Loading REBEL model...\")\n",
    "model_name = \"Babelscape/rebel-large\"\n",
    "tokenizer = AutoTokenizer.from_pretrained(model_name)\n",
    "model = AutoModelForSeq2SeqLM.from_pretrained(model_name)\n",
    "\n",
    "# === Load input JSON data ===\n",
    "with open(\"./Relation Extraction/disease_chromosome_relations.json\", \"r\", encoding=\"utf-8\") as f:\n",
    "    data = json.load(f)\n",
    "\n",
    "# === Prepare inputs for batch processing ===\n",
    "sentences = [f\"<triplet> {item['sentence']}\" for item in data]\n",
    "batch_size = 4  # adjust based on your GPU/CPU capacity\n",
    "\n",
    "# === Extract triplets from REBEL output ===\n",
    "def extract_triplets(text):\n",
    "    triplets = []\n",
    "    if \"<triplet>\" not in text:\n",
    "        return triplets\n",
    "    for segment in text.split(\"<triplet>\")[1:]:\n",
    "        parts = segment.strip().split(\"<subj>\")\n",
    "        for part in parts:\n",
    "            if \"<obj>\" in part:\n",
    "                subject = part.split(\"<obj>\")[0].strip()\n",
    "                rest = part.split(\"<obj>\")[1].strip()\n",
    "                if \"<rel>\" in rest:\n",
    "                    obj, rel = rest.split(\"<rel>\")\n",
    "                    triplets.append({\n",
    "                        \"subject\": subject.strip(),\n",
    "                        \"relation\": rel.strip(),\n",
    "                        \"object\": obj.strip()\n",
    "                    })\n",
    "    return triplets\n",
    "\n",
    "# === Run batched inference and filter ===\n",
    "print(\"üîç Extracting relations from sentences...\")\n",
    "results = []\n",
    "\n",
    "for i in tqdm(range(0, len(sentences), batch_size)):\n",
    "    batch = sentences[i:i + batch_size]\n",
    "    batch_inputs = tokenizer(batch, return_tensors=\"pt\", padding=True, truncation=True)\n",
    "    outputs = model.generate(**batch_inputs, max_length=512, num_beams=4)\n",
    "    decoded_batch = tokenizer.batch_decode(outputs, skip_special_tokens=True)\n",
    "\n",
    "    for j, decoded in enumerate(decoded_batch):\n",
    "        item = data[i + j]\n",
    "        t1 = item[\"target1_word\"].lower()\n",
    "        t2 = item[\"target2_word\"].lower()\n",
    "        triplets = extract_triplets(decoded)\n",
    "\n",
    "        for t in triplets:\n",
    "            subj, obj = t[\"subject\"].lower(), t[\"object\"].lower()\n",
    "            if (t1 in subj and t2 in obj) or (t2 in subj and t1 in obj):\n",
    "                results.append({\n",
    "                    \"pmid\": item[\"pmid\"],\n",
    "                    \"subject\": t[\"subject\"],\n",
    "                    \"relation\": t[\"relation\"],\n",
    "                    \"object\": t[\"object\"],\n",
    "                    \"sentence\": item[\"sentence\"]\n",
    "                })\n",
    "\n",
    "# === Save results to JSON ===\n",
    "with open(\"extracted_relations.json\", \"w\", encoding=\"utf-8\") as f_out:\n",
    "    json.dump(results, f_out, ensure_ascii=False, indent=2)\n",
    "\n",
    "print(f\"\\n‚úÖ Extracted {len(results)} relations. Saved to 'extracted_relations.json'.\")\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90a44921-ef78-4252-a5ae-ca3546c54671",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
